File: src/main/java/org/example/kalkulationsprogramm/service/LieferantArtikelpreisMapper.java
package org.example.kalkulationsprogramm.service;

import org.example.kalkulationsprogramm.domain.LieferantenArtikelPreise;
import org.example.kalkulationsprogramm.dto.Lieferant.LieferantArtikelpreisDto;
import org.springframework.stereotype.Component;

import java.time.ZoneId;
import java.util.Date;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@Component
public class LieferantArtikelpreisMapper {

    public LieferantArtikelpreisDto toDto(LieferantenArtikelPreise entity) {
        if (entity == null) {
            return null;
        }
        LieferantArtikelpreisDto dto = new LieferantArtikelpreisDto();
        if (entity.getArtikel() != null) {
            dto.setArtikelId(entity.getArtikel().getId());
            dto.setProduktname(entity.getArtikel().getProduktname());
            dto.setProdukttext(entity.getArtikel().getProdukttext());
            if (entity.getArtikel().getWerkstoff() != null) {
                dto.setWerkstoff(entity.getArtikel().getWerkstoff().getName());
            }
        }
        dto.setExterneArtikelnummer(entity.getExterneArtikelnummer());
        dto.setPreis(entity.getPreis());
        dto.setPreisAenderungsdatum(toLocalDate(entity.getPreisAenderungsdatum()));
        return dto;
    }

    public List<LieferantArtikelpreisDto> toDtoList(List<LieferantenArtikelPreise> entities) {
        if (entities == null) {
            return List.of();
        }
        return entities.stream()
                .filter(Objects::nonNull)
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    private java.time.LocalDate toLocalDate(Date date) {
        if (date == null) {
            return null;
        }
        return date.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/LieferantArtikelpreisMapper.kt
package org.example.kalkulationsprogramm.service

import org.example.kalkulationsprogramm.domain.Artikel
import org.example.kalkulationsprogramm.domain.LieferantenArtikelPreise
import org.example.kalkulationsprogramm.dto.Lieferant.LieferantArtikelpreisDto
import org.springframework.stereotype.Component
import java.time.ZoneId
import java.util.Date

@Component
class LieferantArtikelpreisMapper {
    fun toDto(entity: LieferantenArtikelPreise?): LieferantArtikelpreisDto? {
        if (entity == null) {
            return null
        }

        val dto = LieferantArtikelpreisDto()
        val artikel = entity.artikel
        if (artikel != null) {
            dto.artikelId = longValue(artikel, "getId")
            dto.produktname = stringValue(artikel, "getProduktname")
            dto.produkttext = stringValue(artikel, "getProdukttext")
            val werkstoff = invokeGetter(artikel, "getWerkstoff")
            if (werkstoff != null) {
                dto.werkstoff = stringValue(werkstoff, "getName")
            }
        }
        dto.externeArtikelnummer = entity.externeArtikelnummer
        dto.preis = entity.preis
        dto.preisAenderungsdatum = toLocalDate(entity.preisAenderungsdatum)
        return dto
    }

    fun toDtoList(entities: List<LieferantenArtikelPreise>?): List<LieferantArtikelpreisDto> =
        entities.orEmpty().mapNotNull { toDto(it) }

    private fun toLocalDate(date: Date?) =
        date?.toInstant()?.atZone(ZoneId.systemDefault())?.toLocalDate()

    private fun longValue(target: Any, getter: String): Long? =
        invokeGetter(target, getter) as? Long

    private fun stringValue(target: Any, getter: String): String? =
        invokeGetter(target, getter) as? String

    private fun invokeGetter(target: Any, getter: String): Any? =
        target.javaClass.getMethod(getter).invoke(target)
}
